pandas/_libs/hashtable_class_helper.pxi.in

```diff
@@ -976,42 +1006,100 @@ cdef class PyObjectHashTable(HashTable):
+                    labels[i] = count
+                    count += 1
+            elif return_inverse:
+                # k falls into a previous bucket
+                # only relevant in case we need to construct the inverse
+                idx = self.table.vals[k]
+                labels[i] = idx
+
+        if return_inverse:
+            return uniques.to_array(), np.asarray(labels)
+        return uniques.to_array()
+
+    def unique(self, ndarray[object] values, bint return_inverse=False):
+        """
+        Calculate unique values and labels (no sorting!)
 
-        return np.asarray(labels)
+        Parameters
+        ----------
+        values : ndarray[object]
+            Array of values of which unique will be calculated
+        return_inverse : boolean, default False
+            Whether the mapping of the original array values to their location
+            in the vector of uniques should be returned.
+
+        Returns
+        -------
+        uniques : ndarray[object]
+            Unique values of input, not sorted
+        labels : ndarray[int64] (if return_inverse)
+            The labels from values to uniques
+        """
+        uniques = ObjectVector()
+        return self._unique(values, uniques, ignore_na=False,
+                            return_inverse=return_inverse)
 
     def factorize(self, ndarray[object] values, Py_ssize_t na_sentinel=-1,
                   object na_value=None):
-        uniques = ObjectVector()
-        labels = self._factorize(values, uniques=uniques,
-                                 na_sentinel=na_sentinel, na_value=na_value)
-        return labels, uniques.to_array()
+        """
+        Calculate unique values and labels (no sorting!)
+
+        Missing values are not included in the "uniques" for this method.
+        The labels for any missing values will be set to "na_sentinel"
+
+        Parameters
+        ----------
+        values : ndarray[object]
+            Array of values of which unique will be calculated
+        na_sentinel : Py_ssize_t, default -1
+            Sentinel value used for all NA-values in inverse
+        na_value : object, default None
+            Value to identify as missing. If na_value is None, then None _plus_
+            any value "val" satisfying val != val is considered missing.
+            If na_value is not None, then _additionally_, any value "val"
+            satisfying val == na_value is considered missing.
+
+        Returns
+        -------
+        labels : ndarray[int64]
+            The labels from values to uniques
+        uniques : ndarray[object]
+            Unique values of input, not sorted
+        """
+        uniques_vector = ObjectVector()
+        uniques, labels = self._unique(values, uniques_vector,
+                                       na_sentinel=na_sentinel,
+                                       na_value=na_value, ignore_na=True,
+                                       return_inverse=True)
```

Comment:
this seems silly, change the new function or change what uses factorize (prob much harder)

Reply:
Well, I could change the cython code easily, but the larger problem is that `.factorize` has this signature already, but then, `np.unique` already appends the inverse in the back, and prepending an output to the signature of `.unique` based on a kwarg is also not great.

I think having that switch around here is the least bad solution, barring a deprecation cycle that switches `.factorize` to return the indices as the second element of the tuple.

Reply:
see above, just the cython signature, we are not going to (nor need to) change the user facing signature